Reduce custom property storage for sites with large numbers of custom properties on the root

heycam opened the pull request:
#### 147e3eca970c1bf4808e592c92d1a94ef2723b2f
<pre>
Reduce custom property storage for sites with large numbers of custom properties on the root
<a href="https://bugs.webkit.org/show_bug.cgi?id=252142">https://bugs.webkit.org/show_bug.cgi?id=252142</a>
&lt;rdar://problem/105372920&gt;

Reviewed by Antti Koivisto and Darin Adler.

Some sites have a large number of custom properties set on the root element,
inheriting into the entire document. When different custom properties or values
are set on descendant elements, we clone all of the inherited properties.

This patch changes the storage in StyleCustomPropertyData so that the
first time we copy() one that has some custom properties in it, we point
the cloned object to the original StyleCustomPropertyData as the source
of inherited custom properties, and leave m_values (now renamed to
m_ownValues) empty. Any subsequent copy() of a StyleCustomPropertyData that
already has a pointer to the inherited custom properties will copy that
pointer, and clone m_ownValues like before.

We don&apos;t generate an unlimited linked list of inherited custom
properties to look up, to avoid custom property lookup time having
worst case linear time in the depth of the tree.

This saves around 800 KB on the youtube.com in Membuster.

* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/CSSComputedStyleDeclaration.cpp:
(WebCore::CSSComputedStyleDeclaration::item const):
* Source/WebCore/css/typedom/ComputedStylePropertyMapReadOnly.cpp:
(WebCore::ComputedStylePropertyMapReadOnly::entries const):
* Source/WebCore/rendering/style/RenderStyle.cpp:
(WebCore::RenderStyle::setCustomPropertyValue):
* Source/WebCore/rendering/style/RenderStyle.h:
(WebCore::RenderStyle::inheritedCustomProperties const):
(WebCore::RenderStyle::nonInheritedCustomProperties const):
* Source/WebCore/rendering/style/StyleCustomPropertyData.cpp: Added.
(WebCore::StyleCustomPropertyData::StyleCustomPropertyData):
(WebCore::StyleCustomPropertyData::get const):
(WebCore::StyleCustomPropertyData::set):
(WebCore::StyleCustomPropertyData::operator== const):
(WebCore::StyleCustomPropertyData::forEach const):
(WebCore::StyleCustomPropertyData::findKeyAtIndex const):
(WebCore::StyleCustomPropertyData::size const):
* Source/WebCore/rendering/style/StyleCustomPropertyData.h:
(WebCore::StyleCustomPropertyData::copy const):
(WebCore::StyleCustomPropertyData::operator!= const):
(WebCore::StyleCustomPropertyData::operator== const): Deleted.
(WebCore::StyleCustomPropertyData::setCustomPropertyValue): Deleted.
(WebCore::StyleCustomPropertyData::StyleCustomPropertyData): Deleted.

Canonical link: <a href="https://commits.webkit.org/260319@main">https://commits.webkit.org/260319@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/63976f536f0601bd474ac3aa8bbdb84c794a4efe

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/107871 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/16926 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/40756 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/117009 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/116361 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/111760 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/18321 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/8242 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/100051 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/113631 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/13826 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/97041 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/41532 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/95734 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/28679 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/83383 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/9858 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/30027 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/10564 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/6922 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/16010 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/49615 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7139 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/12135 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/rendering/style/StyleCustomPropertyData.h:
```diff
@@ -35,36 +36,111 @@ class StyleCustomPropertyData : public RefCounted<StyleCustomPropertyData> {
+inline AtomString StyleCustomPropertyData::keyAtIndex(unsigned index) const
+{
+    unsigned currentIndex = 0;
+    Vector<AtomString> keys;
```
This local seems to be unused.

darinadler commented on Source/WebCore/rendering/style/StyleCustomPropertyData.h:
```diff
@@ -35,36 +36,111 @@ class StyleCustomPropertyData : public RefCounted<StyleCustomPropertyData> {
+{
+    if (m_inheritedValues) {
+        ASSERT(!m_inheritedValues->m_inheritedValues);
+        for (const auto& iter : m_inheritedValues->m_values) {
```
Can this have another name? We frown on abbreviations and I don’t think `iter` is particularly good either. How about `inheritedValue`?

darinadler replied:
Or `entry`.

darinadler replied:
Also no need for the `const`.

darinadler commented on Source/WebCore/rendering/style/StyleCustomPropertyData.h:
```diff
@@ -35,36 +36,111 @@ class StyleCustomPropertyData : public RefCounted<StyleCustomPropertyData> {
 };
 
+inline StyleCustomPropertyData::StyleCustomPropertyData(const StyleCustomPropertyData& other)
+    : RefCounted<StyleCustomPropertyData>()
```
I think this can just be `RefCounted` without the template argument.

darinadler commented on Source/WebCore/rendering/style/StyleCustomPropertyData.h:
```diff
@@ -35,36 +36,111 @@ class StyleCustomPropertyData : public RefCounted<StyleCustomPropertyData> {
+        callback(iter);
+}
+
+inline AtomString StyleCustomPropertyData::keyAtIndex(unsigned index) const
```
Since this is an inefficient algorithm can we use the word “find” or something like it in the function name?

anttijk commented on Source/WebCore/rendering/style/StyleCustomPropertyData.h:
```diff
@@ -35,36 +36,111 @@ class StyleCustomPropertyData : public RefCounted<StyleCustomPropertyData> {
+        ++m_sizeExcludingInheritedValues;
+}
+
+inline bool StyleCustomPropertyData::operator==(const StyleCustomPropertyData& other) const
```
Maybe we should add a .cpp and out-of-line these.

anttijk commented on Source/WebCore/rendering/style/RenderStyle.h:
```diff
@@ -206,8 +206,8 @@ class RenderStyle {
+    const StyleCustomPropertyData& nonInheritedCustomProperties() const { return m_rareNonInheritedData->customProperties.get(); }
     const CSSCustomPropertyValue* customPropertyValue(const AtomString&) const;
 
     void deduplicateCustomProperties(const RenderStyle&);
```
How does this approach play with `deduplicateCustomProperties` which was an earlier attempt to reduce memory consumption here?

heycam replied:
Per offline conversation, we still want to call `deduplicateCustomProperties`, since we still can (and still want to avoid) setting a new `StyleCustomPropertyData` object with equal values.

anttijk commented on Source/WebCore/rendering/style/StyleCustomPropertyData.h:
```diff
@@ -35,36 +36,111 @@ class StyleCustomPropertyData : public RefCounted<StyleCustomPropertyData> {
-    { }
+    StyleCustomPropertyData(const StyleCustomPropertyData&);
+
+    RefPtr<const StyleCustomPropertyData> m_inheritedValues;
```
Is there a danger we are creating cycles here? Note that with MatchedDeclarationsCache we may reuse a RenderStyle on another element because it matched the same style declarations.

I suppose this is prevented by how the copy constructor works?

heycam replied:
Right, the copy constructor will only ever "inherit" from another `StyleCustomPropertyData` if that other object does not inherit from something else.

anttijk commented on Source/WebCore/rendering/style/StyleCustomPropertyData.h:
```diff
@@ -35,36 +36,111 @@ class StyleCustomPropertyData : public RefCounted<StyleCustomPropertyData> {
+inline StyleCustomPropertyData::StyleCustomPropertyData(const StyleCustomPropertyData& other)
+    : RefCounted<StyleCustomPropertyData>()
+    , m_inheritedValues(!other.m_inheritedValues && !other.m_values.isEmpty() ? &other : other.m_inheritedValues.get())
+    , m_values(!other.m_inheritedValues && !other.m_values.isEmpty() ? CustomPropertyValueMap() : other.m_values)
```
m_inheritedValues is initialized her for non-inherited custom properties too. It should have a different name so it doesn't get confused with CSS inheritance. m_parentValue/m_clonedValues or something?

This could use a comment explaining how things work, `!other.m_inheritedValues && !other.m_values.isEmpty()` is bit subtle.

heycam replied:
`m_parentValues` for the reference to the other `StyleCustomPropertyData` object sounds good. Maybe `m_ownValues` for the `HashMap`?

I will factor out that condition into a separate function.

anttijk commented on Source/WebCore/rendering/style/StyleCustomPropertyData.h:
```diff
@@ -35,36 +36,111 @@ class StyleCustomPropertyData : public RefCounted<StyleCustomPropertyData> {
+    StyleCustomPropertyData(const StyleCustomPropertyData&);
+
+    RefPtr<const StyleCustomPropertyData> m_inheritedValues;
+    CustomPropertyValueMap m_values;
```
CustomPropertyValueMap is no longer exposed so maybe the typedef can be removed or made private

anttijk commented on Source/WebCore/rendering/style/StyleCustomPropertyData.h:
```diff
@@ -35,36 +36,111 @@ class StyleCustomPropertyData : public RefCounted<StyleCustomPropertyData> {
+{
+    auto addResult = m_values.set(name, WTFMove(value));
+    if (addResult.isNewEntry && (!m_inheritedValues || !m_inheritedValues->m_values.contains(name)))
+        ++m_sizeExcludingInheritedValues;
```
Similarly some other term than 'inherited' would reduce confusion.

anttijk commented on Source/WebCore/rendering/style/StyleCustomPropertyData.h:
```diff
@@ -35,36 +36,111 @@ class StyleCustomPropertyData : public RefCounted<StyleCustomPropertyData> {
+    return nullptr;
+}
+
+inline void StyleCustomPropertyData::set(const AtomString& name, Ref<const CSSCustomPropertyValue>&& value)
```
This relies in properties never being set after style construction (or you'd mutate the inheriting styles too). Fortunately this should be the case.

heycam replied:
I think the use of `DataRef`s on `StyleRareNonInheritedData` and `StyleRareInheritedData` should prevent this even if we did set property values after style construction. I could change the `RefPtr` here to be a `DataRef` to maybe make that clearer, even though we don't modify property values through that `RefPtr`.

heycam replied:
Oh, I guess `DataRef` doesn't allow a null pointer. I'll leave it as `RefPtr`.

heycam committed:
Reduce custom property storage for sites with large numbers of custom properties on the root
https://bugs.webkit.org/show_bug.cgi?id=252142
<rdar://problem/105372920>

Reviewed by Antti Koivisto and Darin Adler.

Some sites have a large number of custom properties set on the root element,
inheriting into the entire document. When different custom properties or values
are set on descendant elements, we clone all of the inherited properties.

This patch changes the storage in StyleCustomPropertyData so that the
first time we copy() one that has some custom properties in it, we point
the cloned object to the original StyleCustomPropertyData as the source
of inherited custom properties, and leave m_values (now renamed to
m_ownValues) empty. Any subsequent copy() of a StyleCustomPropertyData that
already has a pointer to the inherited custom properties will copy that
pointer, and clone m_ownValues like before.

We don't generate an unlimited linked list of inherited custom
properties to look up, to avoid custom property lookup time having
worst case linear time in the depth of the tree.

This saves around 800 KB on the youtube.com in Membuster.

* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/CSSComputedStyleDeclaration.cpp:
(WebCore::CSSComputedStyleDeclaration::item const):
* Source/WebCore/css/typedom/ComputedStylePropertyMapReadOnly.cpp:
(WebCore::ComputedStylePropertyMapReadOnly::entries const):
* Source/WebCore/rendering/style/RenderStyle.cpp:
(WebCore::RenderStyle::setCustomPropertyValue):
* Source/WebCore/rendering/style/RenderStyle.h:
(WebCore::RenderStyle::inheritedCustomProperties const):
(WebCore::RenderStyle::nonInheritedCustomProperties const):
* Source/WebCore/rendering/style/StyleCustomPropertyData.cpp: Added.
(WebCore::StyleCustomPropertyData::StyleCustomPropertyData):
(WebCore::StyleCustomPropertyData::get const):
(WebCore::StyleCustomPropertyData::set):
(WebCore::StyleCustomPropertyData::operator== const):
(WebCore::StyleCustomPropertyData::forEach const):
(WebCore::StyleCustomPropertyData::findKeyAtIndex const):
(WebCore::StyleCustomPropertyData::size const):
* Source/WebCore/rendering/style/StyleCustomPropertyData.h:
(WebCore::StyleCustomPropertyData::copy const):
(WebCore::StyleCustomPropertyData::operator!= const):
(WebCore::StyleCustomPropertyData::operator== const): Deleted.
(WebCore::StyleCustomPropertyData::setCustomPropertyValue): Deleted.
(WebCore::StyleCustomPropertyData::StyleCustomPropertyData): Deleted.

Canonical link: https://commits.webkit.org/260319@main
